core/src/test/scala/integration/kafka/api/TransactionsTest.scala

```diff
@@ -532,6 +532,55 @@ class TransactionsTest extends KafkaServerTestHarness {
     }
   }
 
+  @Test(expected = classOf[TimeoutException])
+  def testInitTransactionFailWithBadBrokers() {
+    val producer = createTransactionalProducerToConnectNonExistentBrokers()
+
+    try {
+      producer.initTransactions()
+      producer.beginTransaction()
+    } finally {
+      producer.close()
+    }
+  }
+
+  @Test(expected = classOf[KafkaException])
+  def testConsecutivelyRunInitTransactions(): Unit = {
+    val producer = createTransactionalProducer(transactionalId = "normalProducer")
+
+    try {
+      producer.initTransactions()
+      producer.initTransactions()
+      fail("Should have raised a KafkaException")
+    } finally {
+      producer.close()
+    }
+  }
+
+  @Test(expected = classOf[KafkaException])
+  def testOnlyCanExecuteCloseAfterFailInitTransaction(): Unit = {
+    val producer = createTransactionalProducerToConnectNonExistentBrokers()
+    try {
+      producer.initTransactions()
+    } catch {
+      case _: TimeoutException => // expected
+    }
+    // other transactional operations should not be allowed even when we capture the error after initTransaction failed
+    try {
+      producer.beginTransaction()
+    } finally {
+      producer.close() // should successfully close the producer
+    }
+  }
+
+  private def createTransactionalProducerToConnectNonExistentBrokers(): KafkaProducer[Array[Byte], Array[Byte]] = {
+    val props = new Properties()
+    props.put(ProducerConfig.TRANSACTIONAL_ID_CONFIG, "test-transaction-id")
+    val producer = TestUtils.createNewProducer(brokerList = "192.168.1.1:9092", maxBlockMs = 1000,
```

Comment:
Is there a way we can test this without depending on an IP directly like this? I don't think we can assume that the builds will always be sandboxed. I actually think `KafkaProducerTest` might be a better place since it lets us mock the network layer.